pyrogram/client/client.py

```diff
@@ -2306,3 +2309,100 @@ def download_media(self, message: types.Message, file_name: str = None):
         self.download_queue.put((media, file_name, done))
 
         done.wait()
+
+    def add_contacts(self,
+                     phone: Union[int, str] = None,
+                     first_name: str = None,
+                     last_name: str = None,
+                     input_phone_contact_list: List[InputPhoneContact] = None):
+        if (phone is None or first_name is None) and \
+                input_phone_contact_list is None:
+            log.warning("(phone and first_name) or input_phone_contact_list "
+                        "must be not None")
+            return None
+
+        if phone is not None and first_name is not None:
+            if str(phone)[0] != '+':
+                phone = '+' + str(phone)
+            input_phone_contact_list = []
+            input_phone_contact = InputPhoneContact(client_id=0,
+                                                    phone=phone,
+                                                    first_name=first_name,
+                                                    last_name=last_name or '')
+            input_phone_contact_list.append(input_phone_contact)
+
+        # make sure that we send only InputPhoneContact
+        inner_input_phone_contact_list = []
+        for contact in input_phone_contact_list:
+            if isinstance(contact, InputPhoneContact):
+                inner_input_phone_contact_list.append(contact)
+
+        imported_contacts = self.send(
+            functions.contacts.ImportContacts(inner_input_phone_contact_list))
+
+        for user in imported_contacts.users:
+            if isinstance(user, User):
+                if user.id in self.peers_by_id:
+                    continue
+
+                if user.access_hash is None:
+                    continue
+
+                input_peer = InputPeerUser(
+                    user_id=user.id,
+                    access_hash=user.access_hash
+                )
+
+                self.peers_by_id[user.id] = input_peer
+
+                if user.username is not None:
+                    self.peers_by_username[user.username] = input_peer
+
+        return imported_contacts
+
+    def delete_contacts(self, _id: int = None,
+                        ids_list: Union[
+                            List[int], List[InputUser]] = None):
+        if _id is None and ids_list is None:
+            log.warning('id or ids_list must be not None')
+            return False
+
+        contacts = self.get_contacts()
```

Comment:
There's no need to check whether the Contact we want to remove exists in the user's contacts list

Reply:
what if you don't have a conversation with a contact you want to remove, you need to get ```access_hash```

Reply:
if a call to resolve_peer returns nothing we simply ignore that

Reply:
so, in order to make it working for contacts, the lib must also fetch contacts when it starts, just like get_dialogs

Reply:
so, if you don't have a conversation with a contact you cannot remove it? or i didn't get it

Reply:
There's no need to have a conversation with a contact in order to remove it.
Here's why: the raw function GetContacts returns this: https://docs.pyrogram.ml/types/contacts/Contacts, and the `users` lists contains the detailed info about all the contacts, with the `access_hash` too.

Reply:
yeah, that's why i add ```contacts = self.get_contacts()```, to obtain `access_hash`

Reply:
But getting contacts at the start will enable yet another feature, which is sending messages to contacts by using their phone numbers too

Reply:
you mean, you want to add ```self.get_contacts()``` to the ```self.start()```? well, if you accidentally add a contact, you need to perform client restart to be able to delete it, right? or manually get ```access_hash``` from GetContacts

Reply:
Nope, add_contacts will call `fetch_peers` do add the new contact in the internal dictionaries, this is possible because https://docs.pyrogram.ml/types/contacts/ImportedContacts **also** contains the `users` list

Reply:
but if you restart the Client, `self.get_dialogs()` won't add new contacts to peers_by_id .._username. i double checked. could you please also test it. maybe my mistake.

Reply:
i mean in that case if you haven't start a dialog with new contact

Reply:
Well, of course it won't add new contacts, because the code was missing, I've added it just now. Check new commits on this PR please

Reply:
ok, i've understood your idea. thanks!